File: core/src/com/alta_v2/physics/task/rotatePlayer/RotatePersonTask.java
package com.alta_v2.physics.task.rotatePlayer;

import com.alta_v2.physics.executionContext.Tenant;
import com.alta_v2.physics.executionContext.reserveData.ReservablePersonView;
import com.alta_v2.physics.task.MovementDirection;
import com.alta_v2.physics.task.ResultTiledMapTask;
import com.alta_v2.physics.task.TiledMapTask;
import com.alta_v2.physics.task.resultObserver.BaseTaskResultObserver;
import lombok.Getter;

/**
 * Provides the logic related to rotate of player.
 */
public class RotatePersonTask implements ResultTiledMapTask {

    private final MovementDirection direction;
    private final ReservablePersonView personView;
    private final Tenant tenant = new Tenant("rotate-person-task");

    @Getter
    private boolean completed;

    @Getter
    private BaseTaskResultObserver result = new BaseTaskResultObserver();

    public RotatePersonTask(MovementDirection direction, ReservablePersonView personView) {
        this.personView = personView;
        this.direction = direction;
        this.completed = false;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void act(float delta) {
        this.personView
                .reserve(this.tenant)
                .setValue(MovementDirection.getPersonView(direction), this.tenant)
                .release(this.tenant);
        this.completed = true;

        this.result.submitComplete();
    }

    @Override
    public void destroy() {
        this.result.destroy();
        this.result = null;
    }
}

File: core/src/com/alta_v2/physics/task/rotatePlayer/RotatePersonTask.kt
package com.alta_v2.physics.task.rotatePlayer

import com.alta_v2.physics.executionContext.Tenant
import com.alta_v2.physics.executionContext.reserveData.ReservablePersonView
import com.alta_v2.physics.task.MovementDirection
import com.alta_v2.physics.task.ResultTiledMapTask
import com.alta_v2.physics.task.resultObserver.TaskResult

/**
 * Provides the logic related to rotate of player.
 */
class RotatePersonTask(private val direction: MovementDirection,
                       private val personView: ReservablePersonView) : ResultTiledMapTask {
    private val tenant = Tenant("rotate-person-task")

    private var completed = false
    private val taskResult = TaskResult()

    /**
     * {@inheritDoc}
     */
    override fun act(delta: Float) {
        personView
                .reserve(tenant)
                .setValue(MovementDirection.getPersonView(direction), tenant)
                .release(tenant)
        completed = true
        taskResult.complete(null)
    }

    override fun isCompleted(): Boolean = completed
    override fun getResult(): TaskResult = taskResult

    override fun destroy() {}

}